test/flint/fmpz_mod_mat-test.jl

```diff
@@ -579,31 +579,129 @@ end
   @test_throws ErrorException inv(matrix(R, 2, 1, [1, 1]))
 end
 
-@testset "ZZModMatrix.solve" begin
-  Z17, = residue_ring(ZZ, ZZ(17))
-  a = matrix(Z17, [1 2 3; 3 2 1; 0 0 2])
-  b = matrix(Z17, [2 1 0 1; 0 0 0 0; 0 1 2 0])
+@testset "ZZModMatrix.solve over $R with $NFTrait" for (R, NFTrait) in [
+    (residue_ring(ZZ, ZZ(18))[1], AbstractAlgebra.Solve.HowellFormTrait()),
+    (residue_ring(ZZ, ZZ(17))[1], AbstractAlgebra.Solve.LUTrait())
+   ]
+  a = matrix(R, [1 2 3; 3 2 1; 0 0 2])
+
+  @test AbstractAlgebra.Solve.matrix_normal_form_type(R) === AbstractAlgebra.Solve.HowellFormTrait()
```

Comment:
Shouldn't the rhs here be `NFTrait`? And if not, why can this ring have two different traits? Or is there something going on that I don't see?

Reply:
It's Z/nZ, so sometimes there are zero divisors and sometimes it is a field. We (Tommy) decided that we always give it the `HowellFormTrait` to not make the used normal form runtime dependent.
There is still code (via flint) that can only be used in the case where n is prime, so I explicitly call this in this testset.

Reply:
Ah ok. Thanks for the explanation